capa/features/extractors/binexport2/insn.py

```diff
@@ -0,0 +1,414 @@
+# Copyright (C) 2023 Mandiant, Inc. All Rights Reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at: [package root]/LICENSE.txt
+# Unless required by applicable law or agreed to in writing, software distributed under the License
+#  is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and limitations under the License.
+import io
+import logging
+from typing import List, Tuple, Iterator
+
+import pefile
+from elftools.elf.elffile import ELFFile
+
+import capa.features.extractors.helpers
+import capa.features.extractors.strings
+from capa.features.insn import API, Number, Mnemonic, OperandNumber
+from capa.features.common import Bytes, String, Feature, Characteristic
+from capa.features.address import Address, AbsoluteVirtualAddress
+from capa.features.extractors.binexport2 import AnalysisContext, FunctionContext, InstructionContext
+from capa.features.extractors.base_extractor import BBHandle, InsnHandle, FunctionHandle
+from capa.features.extractors.binexport2.binexport2_pb2 import BinExport2
+
+logger = logging.getLogger(__name__)
+
+
+def extract_insn_api_features(fh: FunctionHandle, _bbh: BBHandle, ih: InsnHandle) -> Iterator[Tuple[Feature, Address]]:
+    fhi: FunctionContext = fh.inner
+    ii: InstructionContext = ih.inner
+
+    be2 = fhi.ctx.be2
+    idx = fhi.ctx.idx
+    analysis = fhi.ctx.analysis
+
+    instruction = be2.instruction[ii.instruction_index]
+
+    if not instruction.call_target:
+        return
+
+    for call_target_address in instruction.call_target:
+        if call_target_address in analysis.thunks:
+            vertex_index = analysis.thunks[call_target_address]
+        elif call_target_address not in idx.vertex_index_by_address:
+            continue
+        else:
+            vertex_index = idx.vertex_index_by_address[call_target_address]
+
+        vertex = be2.call_graph.vertex[vertex_index]
+        if not vertex.HasField("mangled_name"):
+            continue
+
+        function_name = vertex.mangled_name
+        if vertex.HasField("library_index"):
+            # TODO: this seems to be incorrect for Ghidra extractor
+            library = be2.library[vertex.library_index]
+            library_name = library.name
```

Comment:
With https://github.com/google/binexport/commit/39f6445c232bb5caf5c4a2a996de91dfa20c48e8, Ghidra exports library names as module indices per the discussion in https://github.com/google/binexport/issues/126.

Reply:
At the risk opening yet another can of worms, I don't quite understand library vs module (library isn't used by IDA at all??) and wonder if the IDA plugin (and subsequently Ghidra) could use further bug fixes/enhancements to better pick between modules and libraries.

Lets see how the discussion over there evolves.

Reply:
My intuition is that library should contain the name of the DLL, while we should use any associated module to namespace the function name, like `System.out.println`.

Reply:
> My intuition is that library should contain the name of the DLL, while we should use any associated module to namespace the function name, like `System.out.println`.

This is my interpretation as well - I'm interested to see this discussed further on the BinExport side.

Reply:
I've been comparing Ghidra's BinExport output with IDA's BinExport output as IDA's output appears to be the most complete, gold standard according to the BinExport developers. This has been a big help in identifying unimplemented and incorrect exports in the Ghidra exporter.

Reply:
It may also be worth comparing with the exporter that you have access to only at work. I think that has a long history and is hopefully mature.

Reply:
I'm trying to avoid too many exporters in the kitchen at this point. The BinExport protobuf is standard but it's entirely up to the exporter what and how information is stored. Unfortunately this means capa may need workarounds for each different exporter depending on how far they stray from expectations. e.g. the current workarounds we have in place for the Ghidra BinExporter's lack of operand expression tree parsing. By working with the BinExport developers to bring Ghidra closer to parity with IDA and focusing capa's processing on the BinExport files generated by these tools we should have good coverage for most of our user's needs.

Reply:
Yup, totally understand. Mostly mentioning it as a resource to research, to see how it models the data (same or different than IDA?). 

One thing that's unfortunate is that the BinExport2 format doesn't have a place to indicate which exporter produced the data. So we have to develop heuristics to guess when to load additional logic. ... or just fix up *all* the exporters along the way :-)